dev/src/bulk-writer.ts

```diff
@@ -0,0 +1,428 @@
+        !bulkBatch.refsInBatch.has(ref)
+      ) {
+        batch = bulkBatch;
+      }
+    }
+
+    if (batch === null) {
+      if (logWarning) {
+        logger(
+          'BulkWriter',
+          null,
+          'Writing to the same document multiple times will slow down BulkWriter. ' +
+            'Write to unique documents in order to maximize throughput.'
+        );
+      }
+      return this.createNewBatch();
+    } else {
+      return batch;
+    }
+  }
+
+  /**
+   * Create a new batch on the BatchQueue and a mapping from the batch to its
+   * result.
+   */
+  private createNewBatch(): BulkCommitBatch {
+    const tag = requestTag();
+    const deferred = new Deferred<BatchWriteResult[]>();
+    const newBatch = {
+      writeBatch: this.firestore.batch(),
+      refsInBatch: new Set(),
+      tag,
+      pending: false,
+      result: deferred,
+    } as BulkCommitBatch;
+
+    this.batchQueue.push(newBatch);
+    this.resultsMap.set(tag, deferred);
+    return newBatch;
+  }
+
+  /**
+   * Return the result of the write at the given batch tag and operation count.
+   */
+  private getResult(tag: string, opCount: number): Promise<WriteResult> {
+    assert(this.resultsMap.has(tag), 'requestTag not found');
+    return this.resultsMap.get(tag)!.promise.then(results => {
+      assert(
+        results.length >= opCount,
+        'opCount invalid' + results.length + opCount
+      );
+      const result = results[opCount - 1];
+      if (result.writeTime) {
+        return new WriteResult(result.writeTime);
+      } else {
+        throw result.status;
+      }
+    });
+  }
+
+  /**
+   * Attempts to send batches starting from the front of the BatchQueue until a
+   * batch cannot be sent.
+   *
+   * After a batch is complete, try sending batches again.
+   */
+  private sendBatches(): void {
+    let counter = 0;
+    for (const batch of this.batchQueue) {
+      // Only send up until the first batch that cannot be sent.
+      if (!this.isBatchSendable(batch)) {
+        break;
+      }
+
+      this.refsInFlight = new Set([...this.refsInFlight, ...batch.refsInBatch]);
+      batch.pending = true;
+
+      batch.writeBatch.bulkCommit().then(async results => {
+        this.broadcastResults(results, batch.tag);
+
+        // Remove references that are no longer in flight.
+        this.refsInFlight = new Set(
+          [...this.refsInFlight].filter(ref => !batch.refsInBatch.has(ref))
+        );
+
+        this.sendBatches();
+      });
+      counter++;
+    }
+
+    // Remove batches that were successfully sent.
+    this.batchQueue = this.batchQueue.slice(counter);
+  }
+
+  /**
+   * Checks that the provided batch does not write to references that are
+   * currently in flight.
+   */
+  private isBatchSendable(batch: BulkCommitBatch): boolean {
+    for (const ref of batch.refsInBatch) {
+      if (this.refsInFlight.has(ref)) {
+        return false;
+      }
+    }
+    return true;
```

Comment:
This can be done more idiomatically through the use of `Array.any()`. i.e...
```
return batch.refsInBatch.some(ref => this.refsInFlight.has(ref));
```